src/icmp.rs

```diff
@@ -0,0 +1,353 @@
+                0x0B => Self::DestinationUnreachable(Unreachable::NetworkUnreachableForTos),
+                0x0C => Self::DestinationUnreachable(Unreachable::HostUnreachableForTos),
+                0x0D => Self::DestinationUnreachable(Unreachable::CommunicationAdministrativelyProhibited),
+                0x0E => Self::DestinationUnreachable(Unreachable::HostPrecedenceViolation),
+                0x0F => Self::DestinationUnreachable(Unreachable::PrecedentCutoffInEffect),
+                _ => Self::Other(raw),
+            },
+            0x04 => match c {
+                0x00 => Self::SourceQuench,
+                _ => Self::Other(raw),
+            },
+            0x05 => match c {
+                0x00 => Self::Redirect(Redirect::Network),
+                0x01 => Self::Redirect(Redirect::Host),
+                0x02 => Self::Redirect(Redirect::TosAndNetwork),
+                0x03 => Self::Redirect(Redirect::TosAndHost),
+                _ => Self::Other(raw),
+            },
+            0x07 => Self::Reserved,
+            0x08 => Self::EchoRequest,
+            0x09 => Self::RouterAdvertisment,
+            0x0A => Self::RouterSolicication,
+            0x0B => match c {
+                0x00 => Self::TimeExceeded(TimeExceeded::TTL),
+                0x01 => Self::TimeExceeded(TimeExceeded::FragmentReassembly),
+                _ => Self::Other(raw),
+            },
+            0x0C => match c {
+                0x00 => Self::ParameterProblem(ParameterProblem::Pointer),
+                0x01 => Self::ParameterProblem(ParameterProblem::MissingRequiredOption),
+                0x02 => Self::ParameterProblem(ParameterProblem::BadLength),
+                _ => Self::Other(raw),
+            },
+            0x0D => Self::Timestamp,
+            0x0E => Self::TimestampReply,
+            0x2A => Self::ExtendedEchoRequest,
+            0x2B => match c {
+                0x00 => Self::ExtendedEchoReply(ExtendedEchoReply::NoError),
+                0x01 => Self::ExtendedEchoReply(ExtendedEchoReply::MalformedQuery),
+                0x02 => Self::ExtendedEchoReply(ExtendedEchoReply::NoSuchInterface),
+                0x03 => Self::ExtendedEchoReply(ExtendedEchoReply::NoSuchTableEntry),
+                0x04 => Self::ExtendedEchoReply(ExtendedEchoReply::MupltipleInterfacesStatisfyQuery),
+                _ => Self::Other(raw),
+            }
+            _ => Self::Other(raw),
+        }
+    }
+}
+
+fn parse_icmp_code(input: &[u8]) -> IResult<&[u8], IcmpCode> {
+    let (input, code) = number::streaming::be_u16(input)?;
+
+    Ok((input, code.into()))
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum IcmpData {
+    Unreachable { nexthop_mtu: u16, header: IPv4Header, packet: [u8; 8] },
+    Redirect {gateway: Ipv4Addr, header: IPv4Header, packet: [u8; 8] },
+    TimeExceeded {header: IPv4Header, packet: [u8; 8] },
```

Comment:
maybe a transparent type for packet ? just an idea no strong feeling

Reply:
something like `struct Packet([u8; 8])` ?

Reply:
I've added #[repr(transparent)] pub struct IcmpPayloadPacket([u8; 8]);